invenio_rdm_migrator/streams/runner.py

```diff
@@ -23,39 +26,46 @@ def _read_config(self, filepath):
+        self.cache_dir.mkdir(parents=True, exist_ok=True)
+
         self.streams = []
         self.cache = {
-            # used to keep track of what Parent IDs we've already inserted in the PIDs table.
-            # {
-            #     '<parent_pid>': {
-            #         'id': <generated_parent_uuid>,
-            #         'version': {
-            #             'latest_index': 'record_index',
-            #             'latest_id': 'record id',
-            #         }
-            # }
-            "parents": ParentsCache(),
-            "records": RecordsCache(),
+            "parents": ParentsCache(filepath=self.cache_dir / "parents.json"),
+            "records": RecordsCache(filepath=self.cache_dir / "records.json"),
             "communities": {},
         }
-        config = self._read_config(config_filepath)
 
-        for definition in stream_definitions:
-            stream_config = config.get(definition.name)
-            if stream_config is not None:
-                # merge cache objects from stream definition config
-                stream_cache = stream_config.get("load", {}).pop("cache", {})
-                self.cache.update(stream_cache)
-                self.streams.append(
-                    Stream(
-                        definition.name,
-                        definition.extract_cls(**stream_config.get("extract", {})),
-                        definition.transform_cls(**stream_config.get("transform", {})),
-                        definition.load_cls(
-                            **stream_config.get("load", {}), cache=self.cache
-                        ),
+        try:
+            for definition in stream_definitions:
+                stream_config = config.get(definition.name)
+                if stream_config is not None:
+                    # merge cache objects from stream definition config
+                    stream_cache = stream_config.get("load", {}).pop("cache", {})
+                    self.cache.update(stream_cache)
+                    self.streams.append(
+                        Stream(
+                            definition.name,
+                            definition.extract_cls(**stream_config.get("extract", {})),
+                            definition.transform_cls(
+                                **stream_config.get("transform", {})
+                            ),
+                            definition.load_cls(
+                                **stream_config.get("load", {}),
+                                cache=self.cache,
+                                tmp_dir=self.tmp_dir,
+                            ),
+                        )
                     )
-                )
+        finally:
+            for name, cache in self.cache.items():
+                if name == "communities":  # FIXME: implement communities cache
+                    continue
+                cache_file = self.cache_dir / f"{name}.json"
+                cache.dump(cache_file)
```

Comment:
Shouldn't this be done inside `.run()` below? I think here we only initialize the streams, but don't run anything.

Reply:
Ufff, yes, totally true. My bad.